### ISSUE TITLE
Format on save

### ISSUE THREAD

**Issue by SamuFinnos** [STweep/STweep#34]
Pretty self explanatory but would be an amazing feature.

**Comment by Barteling** [comment]
I consider this, on the other hand this might also be something which comes with STweep CLI. See : #32

**Comment by GISED-Link** [comment]
This could be a good idea. Since I have a bug on a specific section of my code, I would prefer to have first a flag/decorator that can stop STweep from running on a given set of lines (I haven't found anything on the documentation). So if STweep still has trouble with some lines of code, I could ignore them until the bug is fixed 👍🏻

**Comment by HAHermsen** [comment]
IMHO this is not such a good idea as I have the autosave feature anabled and the project I am working on is currently 1200+ (and growing) objects. I'd rather see STweep either be enabled by the user or by some agent script on commit via webhooks.

**Comment by nbuchele-KAMP** [comment]
I agree with @HAHermsen. Instead, I'd rather see a format on Build. Builds happen with less frequently than saves, so it wouldn't be overly intrusive. You might need to add some sort of flag to track if a file has already been formatted and has not been changed. Then you wouldn't have to waste time formatting files that don't need it.

**Comment by Rednas83** [comment]
I would love to have this feature too.

Why not make it as a configurable option?

Something like:
![image](https://user-images.githubusercontent.com/25219229/192771778-d57d5194-3ea3-4fea-bdfe-a66d8ba23b6f.png)

**Comment by slimeth** [comment]
What's the status on this feature?

**Comment by HAHermsen** [comment]
Guessing from the previous replies it's under consideration for STweep CLI. 

Besides that I strongly disagree with this implementing this option in any user interactive GUI. A user should decide when to apply the STweep rather then run it automatically after each SAVE. This makes no sense and hinders any normal workflow significantly.

**Comment by slimeth** [comment]
@HAHermsen I agree with you that the user should be able to decide if the feature is enabled or not. Certain users might strongly disagree with this feature while others would strongly benefit from a well formatted code base at every commit. Because it can quickly happen that someone forgets to format a file before a commit by accident.

**Comment by HAHermsen** [comment]
If you want to enforce a STweep action after a commit, one should use a webhook. Besides that, a commit is not equal to a save action. Therefore I seriously doubt the usefulness of a "STweep on Save".
But a "STweep on Commit" webhook however is something I'd like to see so that everyone's code at least is styled the same. In order to enable a "STweep on Commit" under CODESYS, I asked for Python API acces in STweep, while STweep CLI for TwinCAT already is capable of being triggered via webooks. Check it out, you probably already can achieve what you want to use. 

I hope this helps.

**Comment by Rednas83** [comment]
In vscode format on save works perfectly. I don't see why it wouldn't work.
![image](https://user-images.githubusercontent.com/25219229/230335323-7d7b6230-e3f6-49f8-9a6b-a78f06beaea6.png)

Adding formatting to your pipeline is of course also a very good practice.

Perhaps it's better to create an issue for vs for this?

**Comment by slimeth** [comment]
What I would like to avoid is that the local and remote repository commits differ after every push which would happen with the proposed webhook. I would like to use the pipeline rather to verify that the code is properly formatted before every merge and otherwise block it until it is resolved. 

The format on save in vs code is great, unfortunately I'm usually using the TwinCAT XAE Shell for the PLC programming. So I would love to see a similar feature in the STweep TwinCAT XAE Shell plugin.

**Comment by Barteling** [comment]
I still have this option on the list,  but i dont expect it on short notice, 
i do know several stweepers using stweep CLI in combination with a pre commit hook. 
See also https://cookncode.com/twincat/2022/04/14/pre-commit.html

**Comment by MEbneter** [comment]
Would like to see that too. (I was send here by your friendly support after asking about 'formatt on save')
@Barteling ,thats a good tip. I will try that out and maybe use it until this feature is out.

**Comment by rschluep-posalux** [comment]
Hi,
We have huge trouble to setup Twincat correctly. It seems that Twincat loves to reset its parameters from time to time. This lead to formatting issues, especially when we don't want to use tabs.
I really need the feature that format on save. My covorkers always push commits that are not formatted. The `format all` function has also issues on big projects when the number of files is more than 30 files.
